python/ray/rllib/agents/ppo/appo_policy.py

```diff
@@ -275,9 +340,71 @@ def add_values_and_logits(policy):
         out[SampleBatch.VF_PREDS] = policy.value_function
     return out
 
+def validate_config(policy, obs_space, action_space, config):
+    assert config["batch_mode"] == "truncate_episodes", \
+        "Must use `truncate_episodes` batch mode with V-trace."
 
-AsyncPPOTFPolicy = VTraceTFPolicy.with_updates(
+
+def choose_optimizer(policy, config):
+    if policy.config["opt_type"] == "adam":
+        return tf.train.AdamOptimizer(policy.cur_lr)
+    else:
+        return tf.train.RMSPropOptimizer(policy.cur_lr, config["decay"],
+                                         config["momentum"], config["epsilon"])
+
+def clip_gradients(policy, optimizer, loss):
+    grads = tf.gradients(loss, policy.var_list)
+    policy.grads, _ = tf.clip_by_global_norm(grads, policy.config["grad_clip"])
+    clipped_grads = list(zip(policy.grads, policy.var_list))
+    return clipped_grads
+    def __init__(self):
+        self.value_function = self.model.value_function()
+        self.var_list = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES,
+                                          tf.get_variable_scope().name)
+
+    def value(self, ob, *args):
+        feed_dict = {
+            self.get_placeholder(SampleBatch.CUR_OBS): [ob],
+            self.model.seq_lens: [1]
+        }
+        assert len(args) == len(self.model.state_in), \
+            (args, self.model.state_in)
+        for k, v in zip(self.model.state_in, args):
+            feed_dict[k] = v
+        vf = self.get_session().run(self.value_function, feed_dict)
+        return vf[0]
+
+class ValueNetworkMixin(object):
+    def __init__(self):
+        self.value_function = self.model.value_function()
+        self.var_list = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES,
+                                          tf.get_variable_scope().name)
+
+    def value(self, ob, *args):
+        feed_dict = {
+            self.get_placeholder(SampleBatch.CUR_OBS): [ob],
+            self.model.seq_lens: [1]
+        }
+        assert len(args) == len(self.model.state_in), \
+            (args, self.model.state_in)
+        for k, v in zip(self.model.state_in, args):
+            feed_dict[k] = v
+        vf = self.get_session().run(self.value_function, feed_dict)
+        return vf[0]
+
+def setup_mixins(policy, obs_space, action_space, config):
```

Comment:
Why not just import from IMPALA these things instead of repeating code?

Reply:
Good point, this will be in the next update.